File: examples/stripe-pull/consumer.py
"""
Pull-mode consumer for qhook queues.
Polls for messages, processes them, and acks/nacks.
No dependencies beyond Python stdlib.
"""
import json
import signal
import sys
import urllib.request
import urllib.error

QHOOK_URL = "http://localhost:8888"
QUEUE = "payments"

running = True


def shutdown(sig, frame):
    global running
    print("\nShutting down...")
    running = False


signal.signal(signal.SIGINT, shutdown)


def poll(wait: int = 10, batch: int = 1) -> list:
    """Long-poll the queue for messages."""
    url = f"{QHOOK_URL}/api/queues/{QUEUE}/messages?wait={wait}s&batch={batch}"
    req = urllib.request.Request(url)
    try:
        with urllib.request.urlopen(req, timeout=wait + 5) as resp:
            data = json.loads(resp.read())
            return data.get("messages", [])
    except urllib.error.URLError as e:
        print(f"[error] poll failed: {e}")
        return []


def ack(ids: list[str]):
    """Acknowledge successfully processed messages."""
    url = f"{QHOOK_URL}/api/queues/{QUEUE}/ack"
    body = json.dumps({"ids": ids}).encode()
    req = urllib.request.Request(url, data=body, method="POST")
    req.add_header("Content-Type", "application/json")
    with urllib.request.urlopen(req) as resp:
        data = json.loads(resp.read())
        print(f"  acked {data.get('acked', 0)} message(s)")


def nack(ids: list[str]):
    """Negative-acknowledge messages (retry or DLQ)."""
    url = f"{QHOOK_URL}/api/queues/{QUEUE}/nack"
    body = json.dumps({"ids": ids}).encode()
    req = urllib.request.Request(url, data=body, method="POST")
    req.add_header("Content-Type", "application/json")
    with urllib.request.urlopen(req) as resp:
        data = json.loads(resp.read())
        print(f"  nacked: retried={data.get('retried', 0)}, dead={data.get('dead', 0)}")


def process(message: dict) -> bool:
    """Process a single message. Returns True on success."""
    event_type = message.get("event_type", "unknown")
    payload = message.get("payload", {})

    if event_type == "checkout.session.completed":
        print(f"[payment] completed: id={payload.get('id')}, "
              f"amount={payload.get('amount_total')}, "
              f"customer={payload.get('customer')}")
        return True

    if event_type == "charge.failed":
        print(f"[charge] failed: id={payload.get('id')}, "
              f"failure={payload.get('failure_message')}")
        return True

    print(f"[unknown] event_type={event_type}")
    return False


if __name__ == "__main__":
    print(f"Polling queue '{QUEUE}' at {QHOOK_URL} (Ctrl+C to stop)")
    while running:
        messages = poll(wait=10, batch=5)
        if not messages:
            continue

        ok_ids = []
        fail_ids = []
        for msg in messages:
            try:
                if process(msg):
                    ok_ids.append(msg["id"])
                else:
                    fail_ids.append(msg["id"])
            except Exception as e:
                print(f"  [error] {e}")
                fail_ids.append(msg["id"])

        if ok_ids:
            ack(ok_ids)
        if fail_ids:
            nack(fail_ids)

File: examples/stripe-pull/consumer.ts
/**
 * Pull-mode consumer for qhook queues.
 * Polls for messages, processes them, and acks/nacks.
 * Requires Node.js 18+ (built-in fetch).
 */

const QHOOK_URL = "http://localhost:8888";
const QUEUE = "payments";

let running = true;

process.on("SIGINT", () => {
  console.log("\nShutting down...");
  running = false;
});

interface QueueMessage {
  id: string;
  event_id: string;
  event_type: string;
  payload: Record<string, unknown>;
  headers: Record<string, string> | null;
  attempt: number;
  created_at: string;
}

async function poll(wait = 10, batch = 1): Promise<QueueMessage[]> {
  try {
    const resp = await fetch(
      `${QHOOK_URL}/api/queues/${QUEUE}/messages?wait=${wait}s&batch=${batch}`,
      { signal: AbortSignal.timeout((wait + 5) * 1000) },
    );
    const data = (await resp.json()) as { messages: QueueMessage[] };
    return data.messages ?? [];
  } catch (err) {
    console.error(`[error] poll failed: ${err}`);
    return [];
  }
}

async function ack(ids: string[]) {
  const resp = await fetch(`${QHOOK_URL}/api/queues/${QUEUE}/ack`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ids }),
  });
  const data = (await resp.json()) as { acked: number };
  console.log(`  acked ${data.acked} message(s)`);
}

async function nack(ids: string[]) {
  const resp = await fetch(`${QHOOK_URL}/api/queues/${QUEUE}/nack`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ ids }),
  });
  const data = (await resp.json()) as { retried: number; dead: number };
  console.log(`  nacked: retried=${data.retried}, dead=${data.dead}`);
}

function processMessage(msg: QueueMessage): boolean {
  const { event_type, payload } = msg;

  if (event_type === "checkout.session.completed") {
    console.log(
      `[payment] completed: id=${payload.id}, amount=${payload.amount_total}, customer=${payload.customer}`,
    );
    return true;
  }

  if (event_type === "charge.failed") {
    console.log(
      `[charge] failed: id=${payload.id}, failure=${payload.failure_message}`,
    );
    return true;
  }

  console.log(`[unknown] event_type=${event_type}`);
  return false;
}

async function main() {
  console.log(`Polling queue '${QUEUE}' at ${QHOOK_URL} (Ctrl+C to stop)`);

  while (running) {
    const messages = await poll(10, 5);
    if (messages.length === 0) continue;

    const okIds: string[] = [];
    const failIds: string[] = [];

    for (const msg of messages) {
      try {
        if (processMessage(msg)) {
          okIds.push(msg.id);
        } else {
          failIds.push(msg.id);
        }
      } catch (err) {
        console.error(`  [error] ${err}`);
        failIds.push(msg.id);
      }
    }

    if (okIds.length > 0) await ack(okIds);
    if (failIds.length > 0) await nack(failIds);
  }
}

main();
